src/utils/isNodeProcess.ts

```diff
@@ -2,5 +2,5 @@
  * Returns a boolean indicating if the current process is running in NodeJS environment.
  */
 export function isNodeProcess() {
-  return Object.prototype.toString.call(global.process) === '[object process]'
```

Comment:
Can we use optional chaining here?

```js
global?.process
```

The reason why `global.process` is used and its prototype is stringified, is because not any object is necessarily a `process` object. That is to prevent false positives, when you have a global `process` object in your environment, but it's not Node.

Reply:
Thanks for the reply. But I think it may also cause `global is not defined error` ,because typescript will compile `global?.process` to `global === null || global === void 0 ? void 0 : global.process` .In browser, `global` is still not defined. So I think Change to `typeof global === 'object' && typeof global.process==='object'` may be better, what do you think about it?

Reply:
That's a good point. We also need to take into account a React Native process, where, I believe, the `global` is set, but it doesn't have the `globlal.process` property (#203). 

```js
function isNodeProcess() {
  if (typeof global === 'undefined') {
    // Assuming that in React Native `global` exists
    return false
  }

  if (Object.prototype.toString.call(global.process) === '[object process]') {
    // Regular NodeJS process
    return true
  }

  if (SOME_REACT_NATIVE_CONDITION) {
    // We should treat a React Native environment similar to Node,
    // because you cannot setup a Service Worker in React Native,
    // just as you can't in Node.
    return true
  }
}
```

Reply:
Thanks for the reply. I noticed it seems `global.process`  is also be set in React Native process. [Here](https://github.com/facebook/react-native/blob/master/Libraries/Core/setUpGlobals.js) can see its definition. I agree that react native environment should be treated similarly to Node because the service worker is not available. Also, React Native implements XMLHttpRequest and fetch in its environment. Its fetch is implemented by `whatwg-fetch` polyfill, which is also based on XMLHttpRequest. I noticed `node-request-interceptor` intercepts requests using XMLHttpRequest. So it should also work in the React Native environment. But I think there is still one problem that `node-request-interceptor` has used `http` core module, [which is not supported by React Native](https://github.com/facebook/react-native/issues/15048). Although it can be patched by other packages, I think providing a function alone used for React Native may be better, such as setUpReactNativeServer? And may only need to intercept XMLHttpRequest.

Reply:
That's a fair observation regarding the `http` module. I don't think that's a sufficient reason to ship a separate API, but what we may do is check whether `http`/`https` modules exist, and attach the interception only if they do (as a part of `node-request-interceptor` library).

Reply:
Yes, I agree with you. So if check whether `http`/`https` module exist, should we distinguish React Native and node process in `isNodeProcess` function? I think if we check `http`/`https` module, then check `global` and `global.process` and treat React Native and node process similarly may be better. For example, like this
```
function isNodeProcess() {
    return typeof global === 'object' && typeof global.process==='object'
}
```
What do you think of it?

Reply:
Yeah, looks good. I'd keep the prototype check, so an arbitrary object won't be treated as a `process`, even if it shares the namespace:

```js
function isNodeProcess() {
  return typeof global === 'object' && Object.prototype.toString.call(global.process) === '[object process]'
}
```

On the related note, I had some reports that `globals` is not defined in React Native (thanks for the source reference shared previously!). I'd like to approach that with caution. Do you have a chance to test a React Native project to see how this `isNodeProcess` function behaves there? It's expected to return `true` for React Native.

Reply:
I try `console.log(global.process)` in React Native using `expo`, It shows 
```
Finished building JavaScript bundle in 40915ms.
Running application on sdk_gphone_x86.
Object {
  "env": Object {
    "NODE_ENV": "development",
  },
}
```
But `Object.prototype.toString.call(global.process)` shows `[object Object]`, thus Object.prototype.toString.call(global.process) === '[object process]' will be broken.
I noticed [here](https://github.com/facebook/react-native/issues/1331), navigator.product was set to 'ReactNative'. Also [here](https://github.com/facebook/react-native/blob/master/Libraries/Core/setUpNavigator.js) can see its definition. I also try `console.log(navigator.product) `, it shows `ReactNative`.
So I suggest this, just like you mentioned before
```
function isNodeProcess() {
  if (typeof global === 'undefined') {
    // Assuming that in React Native `global` exists
    return false
  }

  if (Object.prototype.toString.call(global.process) === '[object process]') {
    // Regular NodeJS process
    return true
  }

  if (navigator.product==='ReactNative') {
    // We should treat a React Native environment similar to Node,
    // because you cannot setup a Service Worker in React Native,
    // just as you can't in Node.
    return true
  }
}

```
What do you think about it?

Reply:
That's yet again a great find. I like relying on `navigator.product` 👍 

I'd only suggest to unify last two conditions, as they both are treated the same:

```js
if (typeof global !== 'object') {
  return false
}

if (
  Object.prototype.toString.call(global.process) === '[object process]' ||
  navigator.product==='ReactNative'
) {
  return true
}
```

- checking if `typeof global` is not an `object`, to prevent global variable `global` from being treated as an env's global.
- unifying last two conditions.

Looks good to you?

Reply:
Yes, Looks good to me. I have changed the code like that. Thanks for the reply!